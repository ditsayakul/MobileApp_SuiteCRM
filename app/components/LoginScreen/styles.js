import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create({
	
  container: {
		flex: 1,
    backgroundColor: "white",
	},

  logoWrapper: {
    	flex: 0.3,
  },
	
	inputWrapper: {
		flex: 0.2,
    padding: 20,
    	//backgroundColor: 'black'
  },

	statusWrapper: {
		flex: 0.1,
		flexDirection: 'column',
		justifyContent: 'center',
		alignItems: 'center',
		padding: 10,
	},

  buttonWrapper: {
		flex: 0.1,
    padding: 20,
  },

  iconWrap: {
    	paddingHorizontal: 5,
    	alignItems: 'center',
    	justifyContent: 'center',
  },

  inputLineWrap: {
    	flexDirection: 'row',
    	marginVertical: 10,
    	height: 40,
    	borderBottomWidth: 1,
    	borderBottomColor: '#CCCC',
  },

  logo: {
	    flex: 1,
    	width: null,
    	height: null,
  },

  icon: {
    	height: 20,
    	width: 20,
  },

  input: {
    	flex: 1,
    	paddingHorizontal: 10,
  },

});