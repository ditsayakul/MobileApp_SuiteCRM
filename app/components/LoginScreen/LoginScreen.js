import React, { Component } from 'react';
import { StackNavigator } from 'react-navigation';
import { Text, TextInput, Image, View, ScrollView, Button, ActivityIndicator, AsyncStorage, Keyboard } from 'react-native';
import { default as Icon  } from 'react-native-vector-icons/MaterialCommunityIcons';
import { ThemeProvider, Checkbox } from 'react-native-material-ui';
import { styles as defaultStyles } from '../../layout/styles.js'
import { styles, images } from './index.js'
import * as constants from '../../config/const.js'
import { restCall } from '../../core/rest_api.js'


var DEBUG = false;
var MD5 = require("crypto-js/md5");

export class LoginScreen extends Component {

  static navigationOptions = {
    title: 'Exelcia CRM Prospect Manager',
  };

  constructor(props) {
    super(props);
    this.navigate = this.navigate.bind(this);

    this.state={
        status: '',
        session: null,
        isFetching: false,
        ip: null,
        login: null,
        password: null,
        isChecked: false,
        isKeyboardOpen: false,
    };
  }

  navigate(screen){

    var params = {
          session: this.state.session,
          ip: this.state.ip,
    };
    // Navigate to the given screen.
    this.props.navigation.navigate(screen, params);
  }

  connect(){

    this.setInfo(this.state.isChecked);
    this.authentify(this.state.ip, this.state.login, MD5(this.state.password));
    
    if(DEBUG){
      console.log("(LoginScreen) parameters: ");
      console.log("ip = " + this.state.ip);
      console.log("login = " + this.state.login);
      console.log("passwd = " + this.state.password);
      console.log("(LoginScreen) credential (MD5): ");
      console.log(MD5(this.state.password));
    }
  }

  authentify(ip, login, password){

    var credential = '{"user_auth":{"user_name":"'+ login +'","password":"'+ password.toString() +'"}}';

    
    this.setState({isFetching: true});
    
    var onSuccess = function(responseData){
      // We received a response from the server.
      this.setState({isFetching: false, session: responseData.id});
      // Got a session.
      if(this.state.session){
        fc = this.navigate.bind(this);
        fc(constants.listScreen); 
      } 
      // Wrong credential.
      else {
        this.setState({status: 'Mauvais identifiants', session: null});
      }
    }

    var onFailure = function(error){
        // No response from the server or an error.
        this.setState({isFetching: false, status: "Serveur injoignable", session: null});
    }

    restCall("login", credential, this.state.ip, onSuccess.bind(this), onFailure.bind(this));
  }

  onCheckboxCheck(checked, value){
    this.setState({isChecked: checked})
    this.setInfo(checked);
  }

  /**
   * Store/Delete the login and the server IP.
   * If isSaving is true then the data are saved otherwise it is removed.
   **/
  async setInfo(isSaving){
    try {

      if(isSaving){
         await AsyncStorage.setItem('ip', this.state.ip);
         await AsyncStorage.setItem('login', this.state.login);
      } else {
         await AsyncStorage.removeItem('ip');
         await AsyncStorage.removeItem('login');        
      }
    } catch (error) {
        console.log("AsyncStorage error: "+ error);
    }
  }

  // Get the stored values of the server IP and login if there is any.
  async getInfo(key){
    try {
        var value = await AsyncStorage.getItem(key);  
        return value;
    } catch (error) {
        console.log("AsyncStorage error: "+ error);
        return null;
    }
  }

  async componentWillMount(){
      // Auto fill the input field if the user has checked the remind info checkbox.
      var ip = await this.getInfo("ip");
      var login = await this.getInfo("login");
      this.setState({ip: ip ? ip : '', login: login ? login : '', isChecked: (ip || login) ? true : false});
      // Add keyboard listener.
      this.keyboardDidShowListener = Keyboard.addListener('keyboardDidShow', this._keyboardDidShow.bind(this));
      this.keyboardDidHideListener = Keyboard.addListener('keyboardDidHide', this._keyboardDidHide.bind(this));
  }

  componentWillUnmount(){
      this.keyboardDidShowListener.remove();
      this.keyboardDidHideListener.remove();
  }

  _keyboardDidShow(){
    this.setState({isKeyboardOpen: true});
  }
  _keyboardDidHide(){
   this.setState({isKeyboardOpen: false});
  }

  render() {
    return (
      <ThemeProvider uiTheme={constants.uiTheme}>
            <View style={styles.container}>

                {!this.state.isKeyboardOpen && // We dont display the logo when the keyboard is open
                    <View style={styles.logoWrapper}>
                        <Image source={images.logoExelcia} style={styles.logo} resizeMode="contain" />
                    </View>
                }
                <View style={styles.inputWrapper}>
                <ScrollView>
                    {/* Server field*/}
                    <InputLabelRow icon='server-network' value={this.state.ip} placeholder="IP du serveur" isSecure={false}
                    onChangeText={(text) => this.setState({ip: text})} keyboardType='numeric'/>
                    {/* Login field*/}
                    <InputLabelRow icon='account-outline' value={this.state.login} placeholder="Nom d'utilisateur" isSecure={false}
                    onChangeText={(text) => this.setState({login: text})} />
                    {/* Password field*/}
                    <InputLabelRow icon='lock-outline' value={this.state.password} placeholder="Mot de passe" isSecure={true}
                    onChangeText={(text) => this.setState({password: text})} onSubmitEditing={() =>this.connect()}/>
                    <Checkbox  style={{height:40}}value="ok" checked={this.state.isChecked} onCheck={this.onCheckboxCheck.bind(this)} label="Retenir le login et l'adresse du serveur" />
                </ScrollView>
                </View>
                <View style={styles.statusWrapper}>
                    {this.state.isFetching &&
                        <ActivityIndicator style={styles.statusWrapper} size="large" /> ||
                        <Text style={defaultStyles.fontBasicError}> { this.state.status } </Text>
                    }
                </View>
                <View style={styles.buttonWrapper}>
                    <Button
                      onPress={() => this.connect()}
                      title="Connexion"
                      color={constants.uiTheme.palette.primaryColor}
                      disabled={this.state.isFetching}
                    />
                </View>
            </View>
      </ThemeProvider>
    );
  }
}

var InputLabelRow = React.createClass({
    render() {
      return (
        <View style={styles.inputLineWrap}>
            <View style={styles.iconWrap}>
                  <Icon name={this.props.icon} size={30}/>
            </View>
                <TextInput 
                    keyboardType={this.props.keyboardType ? this.props.keyboardType : 'default'}
                    secureTextEntry={this.props.isSecure}
                    onChangeText={this.props.onChangeText}
                    onSubmitEditing={this.props.onSubmitEditing}
                    value={this.props.value}
                    placeholder={this.props.placeholder} 
                    placeholderTextColor="#CCC"
                    style={styles.input} 
                />
        </View>
      );
    }
  });