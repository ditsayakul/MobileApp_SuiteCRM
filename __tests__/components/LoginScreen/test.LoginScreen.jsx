import 'react-native';
import React from 'react';
import { LoginScreen } from '../../../app/components/LoginScreen/LoginScreen'

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';


describe('LoginScreen', () => {
    it('renders without crashing', () => {
       const tree = renderer.create(
       	<LoginScreen />
       	);
    });
});