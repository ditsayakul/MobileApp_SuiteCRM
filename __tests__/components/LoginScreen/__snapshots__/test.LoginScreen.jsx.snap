// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot testing renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 0.5,
      }
    }
  >
    <Image
      resizeMode="contain"
      source={1}
      style={
        Object {
          "flex": 1,
          "height": null,
          "width": null,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "flex": 0.6,
        "padding": 20,
      }
    }
  >
    <RCTScrollView>
      <View>
        <View
          style={
            Object {
              "borderBottomColor": "#CCCC",
              "flex": 1,
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "paddingHorizontal": 5,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={false}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": undefined,
                    "fontSize": 30,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                ]
              }
            >
              
            </Text>
          </View>
          <TextInput
            keyboardType="numeric"
            onChangeText={[Function]}
            onSubmitEditing={undefined}
            placeholder="IP du serveur"
            placeholderTextColor="#CCC"
            secureTextEntry={false}
            style={
              Object {
                "flex": 1,
                "paddingHorizontal": 10,
              }
            }
            value={null}
          />
        </View>
        <View
          style={
            Object {
              "borderBottomColor": "#CCCC",
              "flex": 1,
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "paddingHorizontal": 5,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={false}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": undefined,
                    "fontSize": 30,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                ]
              }
            >
              
            </Text>
          </View>
          <TextInput
            keyboardType="default"
            onChangeText={[Function]}
            onSubmitEditing={undefined}
            placeholder="Nom d'utilisateur"
            placeholderTextColor="#CCC"
            secureTextEntry={false}
            style={
              Object {
                "flex": 1,
                "paddingHorizontal": 10,
              }
            }
            value={null}
          />
        </View>
        <View
          style={
            Object {
              "borderBottomColor": "#CCCC",
              "flex": 1,
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "paddingHorizontal": 5,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={false}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": undefined,
                    "fontSize": 30,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                ]
              }
            >
              
            </Text>
          </View>
          <TextInput
            keyboardType="default"
            onChangeText={[Function]}
            onSubmitEditing={[Function]}
            placeholder="Mot de passe"
            placeholderTextColor="#CCC"
            secureTextEntry={true}
            style={
              Object {
                "flex": 1,
                "paddingHorizontal": 10,
              }
            }
            value={null}
          />
        </View>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
              },
              undefined,
              undefined,
            ]
          }
          testID={undefined}
        >
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            hitSlop={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={undefined}
            testID={undefined}
          >
            <View
              style={
                Object {
                  "backgroundColor": "rgb(0, 0, 0)",
                  "borderRadius": 21.6,
                  "height": 43.2,
                  "left": 2.3999999999999986,
                  "opacity": 0.16,
                  "position": "absolute",
                  "top": 2.3999999999999986,
                  "transform": Array [
                    Object {
                      "scale": 0.01,
                    },
                  ],
                  "width": 43.2,
                  "zIndex": 1,
                }
              }
            />
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "height": 48,
                    "justifyContent": "center",
                    "width": 48,
                  },
                  Object {
                    "height": 48,
                    "width": 48,
                  },
                  undefined,
                ]
              }
            >
              <Text
                accessible={true}
                allowFontScaling={false}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontSize": 24,
                    },
                    undefined,
                    Object {
                      "fontFamily": "Material Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                  ]
                }
              >
                
              </Text>
            </View>
          </View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#000000",
                  "flex": 1,
                  "marginLeft": 20,
                },
                undefined,
                undefined,
                false,
              ]
            }
          >
            Retenir le login et l'adresse du serveur
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 0.1,
        "flexDirection": "column",
        "justifyContent": "center",
        "padding": 10,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "backgroundColor": "white",
          "color": "red",
          "fontSize": 15,
        }
      }
    >
       
      
       
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 0.1,
        "padding": 20,
      }
    }
  >
    <View
      accessibilityComponentType="button"
      accessibilityLabel={undefined}
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#0C42FD",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
              Object {
                "color": "#1F94B7",
              },
            ]
          }
        >
          Connexion
        </Text>
      </View>
    </View>
  </View>
</View>
`;
